src/parsing.jl

```diff
@@ -0,0 +1,260 @@
+    end
+end
+
+function PQValue(jl_result::Result, row::Integer, col::Integer)
+    oid = libpq_c.PQftype(jl_result.result, col - 1)
+
+    return PQValue{oid}(jl_result, row, col)
+end
+
+# strlen
+num_bytes(pqv::PQValue) = libpq_c.PQgetlength(pqv.jl_result.result, pqv.row, pqv.col)
+
+data_pointer(pqv::PQValue) = libpq_c.PQgetvalue(pqv.jl_result.result, pqv.row, pqv.col)
+
+function Base.unsafe_string(pqv::PQValue)
+    return unsafe_string(data_pointer(pqv), num_bytes(pqv))
+end
+
+function string_view(pqv::PQValue)
+    return String(unsafe_wrap(Vector{UInt8}, data_pointer(pqv), num_bytes(pqv)))
+end
+
+# includes null
+bytes_view(pqv::PQValue) = unsafe_wrap(Vector{UInt8}, data_pointer(pqv), num_bytes(pqv) + 1)
+
+Base.String(pqv::PQValue) = unsafe_string(pqv)
+Base.parse(::Type{String}, pqv::PQValue) = unsafe_string(pqv)
+Base.convert(::Type{String}, pqv::PQValue) = String(pqv)
+Base.length(pqv::PQValue) = length(string_view(pqv))
+Base.endof(pqv::PQValue) = endof(string_view(pqv))
+
+# fallback, because Base is bad with string iteration
+Base.parse(::Type{T}, pqv::PQValue) where {T} = parse(T, string_view(pqv))
+
+function Base.start(pqv::PQValue)
+    sv = string_view(pqv)
+    return (sv, start(sv))
+end
+
+function Base.next(pqv::PQValue, state)
+    sv, sv_state = state
+    c, new_sv_state = next(sv, sv_state)
+    return c, (sv, new_sv_state)
+end
+
+function Base.done(pqv::PQValue, state)
+    sv, sv_state = state
+    return done(sv, sv_state)
+end
+
+## integers
+_DEFAULT_TYPE_MAP[:int2] = Int16
+_DEFAULT_TYPE_MAP[:int4] = Int32
+_DEFAULT_TYPE_MAP[:int8] = Int64
+
+## floating point
+_DEFAULT_TYPE_MAP[:float4] = Float32
+_DEFAULT_TYPE_MAP[:float8] = Float64
+
+## oid
+_DEFAULT_TYPE_MAP[:oid] = Oid
+
+## numeric
+_DEFAULT_TYPE_MAP[:numeric] = Decimal
+
+# no default for monetary; needs lconv and lc_monetary from result/connection
+
+## character
+# bpchar is char(n)
+Base.parse(::Type{String}, pqv::PQValue{PQ_SYSTEM_TYPES[:bpchar]}) = rstrip(pqv, ' ')
+# char is "char"
+_DEFAULT_TYPE_MAP[:char] = PQChar
+Base.parse(::Type{PQChar}, pqv::PQValue{PQ_SYSTEM_TYPES[:char]}) = PQChar(first(pqv))
+# varchar, text, and name are all String
+
+## binary data
+
+_DEFAULT_TYPE_MAP[:bytea] = Vector{UInt8}
+function Base.parse(::Type{Vector{UInt8}}, pqv::PQValue{PQ_SYSTEM_TYPES[:bytea]})
+    byte_length = Ref{Csize_t}(0)
+    bytes = bytes_view(pqv)
+
+    unescaped_ptr = libpq_c.PQunescapeBytea(bytes, byte_length)
+
+    if unescaped_ptr == C_NULL
+        error("Could not unescape byte sequence $(String(bytes))")
+    end
+
+    unescaped_vec = copy(unsafe_wrap(Vector{UInt8}, unescaped_ptr, byte_length[]))
+
+    libpq_c.PQfreemem(unescaped_ptr)
+
+    return unescaped_vec
+end
+
+## bool
+# TODO: check whether we ever need this or if PostgreSQL always gives t or f
+_DEFAULT_TYPE_MAP[:bool] = Bool
+const BOOL_TRUE = r"^\s*(t|true|y|yes|on|1)\s*$"i
+const BOOL_FALSE = r"^\s*(f|false|n|no|off|0)\s*$"i
```

Comment:
Can there really be surrounding whitespace?

Reply:
Yep:
```
postgres=# SELECT '   true
'::Bool;
 bool
------
 t
(1 row)
```

However, as the TODO states, I don't know if it's possible for PostgreSQL to return a bool representation other than the default `t`/`f`.